src/data_store.rs

```diff
@@ -0,0 +1,274 @@
+// This file is Copyright its original authors, visible in version control history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. You may not use this file except in
+// accordance with one or both of these licenses.
+
+use crate::logger::{log_error, LdkLogger};
+use crate::types::DynStore;
+use crate::Error;
+
+use lightning::util::ser::{Readable, Writeable};
+
+use std::collections::hash_map;
+use std::collections::HashMap;
+use std::ops::Deref;
+use std::sync::{Arc, Mutex};
+
+pub(crate) trait StorableObjectId: std::hash::Hash + PartialEq + Eq {
+	fn encode_to_hex_str(&self) -> String;
+}
+
+pub(crate) trait StorableObjectUpdate<SO: StorableObject> {
+	fn id(&self) -> SO::Id;
+}
+
+pub(crate) trait StorableObject: Clone + Readable + Writeable {
+	type Id: StorableObjectId;
+	type Update: StorableObjectUpdate<Self>;
+
+	fn id(&self) -> Self::Id;
+	fn update(&mut self, update: &Self::Update) -> bool;
+	fn into_update(&self) -> Self::Update;
+}
+
+#[derive(PartialEq, Eq, Debug, Clone, Copy)]
+pub(crate) enum DataStoreUpdateResult {
+	Updated,
+	Unchanged,
+	NotFound,
+}
+
+pub(crate) struct DataStore<SO: StorableObject, L: Deref>
+where
+	L::Target: LdkLogger,
+{
+	objects: Mutex<HashMap<SO::Id, SO>>,
+	primary_namespace: String,
+	secondary_namespace: String,
+	kv_store: Arc<DynStore>,
+	logger: L,
+}
+
+impl<SO: StorableObject, L: Deref> DataStore<SO, L>
+where
+	L::Target: LdkLogger,
+{
+	pub(crate) fn new(
+		objects: Vec<SO>, primary_namespace: String, secondary_namespace: String,
+		kv_store: Arc<DynStore>, logger: L,
+	) -> Self {
+		let objects =
+			Mutex::new(HashMap::from_iter(objects.into_iter().map(|obj| (obj.id(), obj))));
+		Self { objects, primary_namespace, secondary_namespace, kv_store, logger }
+	}
+
+	pub(crate) fn insert(&self, object: SO) -> Result<bool, Error> {
+		let mut locked_objects = self.objects.lock().unwrap();
+
+		self.persist(&object)?;
+		let updated = locked_objects.insert(object.id(), object).is_some();
```

Comment:
Does the order matter here? The opposite order is used in `insert_or_update`, so there we could have the object in memory but fail persisting.